src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -702,26 +788,94 @@ impl PredicatePushdown for LogicalJoin {
+        let mut right = self.right.predicate_pushdown(right_predicate);
+        let mut output_indices = self.output_indices().to_owned();
+        // This code block pushes the calculation of inputs of join's comparison condition to join's
+        // two sides.
+        {
+            let exprs = on.conjunctions;
+            let (left_exprs, right_exprs, indices_and_ty_of_func_calls) =
+                Self::find_comparison_exprs(left_col_num, right_col_num, &exprs);
+
+            // used to shift indices of input_refs pointing the right side of `join` with
+            // `left_exprs.len`.
+            let mut col_index_mapping = {
+                let map = (0..left_col_num)
+                    .chain(
+                        (left_col_num..left_col_num + right_col_num).map(|i| i + left_exprs.len()),
+                    )
+                    .map(Some)
+                    .collect_vec();
+                ColIndexMapping::new(map)
+            };
+            let (mut exprs, new_output_indices) =
+                Self::remap_exprs_and_output_indices(exprs, output_indices, &mut col_index_mapping);
+            output_indices = new_output_indices;
+
+            // ```ignore
+            // the internal table of join has has the following schema:
+            // original left's columns | left_exprs | original right's columns | right_exprs
+            //```
+            // `left_index` and `right_index` will scan through `left_exprs` and `right_exprs`
+            // respectively.
+            let mut left_index = left_col_num;
+            let mut right_index = left_col_num + left_exprs.len() + right_col_num;
+            // replace chosen function calls.
+            for (i, (index_of_func_call, ty)) in
+                indices_and_ty_of_func_calls.into_iter().enumerate()
+            {
+                let left_input = InputRef::new(left_index, left_exprs[i].return_type());
+                let right_input = InputRef::new(right_index, right_exprs[i].return_type());
+                exprs[index_of_func_call] =
+                    FunctionCall::new(ty, vec![left_input.into(), right_input.into()])
+                        .unwrap()
+                        .into();
+                left_index += 1;
+                right_index += 1;
+            }
+            on = Condition {
+                conjunctions: exprs,
+            };
+
+            // add project to do the calculation.
+            let new_input = |input: PlanRef, appended_exprs: Vec<ExprImpl>| {
+                let mut exprs = input
+                    .schema()
+                    .data_types()
+                    .into_iter()
+                    .enumerate()
+                    .map(|(i, data_type)| InputRef::new(i, data_type).into())
+                    .collect_vec();
+                exprs.extend(appended_exprs);
+                LogicalProject::create(input, exprs)
+            };
+            if !left_exprs.is_empty() {
```

Comment:
The right_exprs is_empty case should be handled separately?

Reply:
no. because left_exprs and right_exprs are of the same length, we just need to check left_exprs.is_empty() here.